massa-execution-worker/src/speculative_ledger.rs

```diff
@@ -360,6 +361,29 @@ impl SpeculativeLedger {
         Ok(())
     }
 
+    /// Gets a copy of a datastore keys for a given address
+    ///
+    /// # Arguments
+    /// * `addr`: address to query
+    ///
+    /// # Returns
+    /// `Some(Vec<Vec<u8>>)` for found keys, `None` if the address does not exist.
+    pub fn get_keys(&self, addr: &Address) -> Option<Vec<Vec<u8>>> {
+
+        let mut added_keys = self.added_changes.get_keys(addr);
+        let (mut final_keys, mut candidate_keys) = get_final_and_candidate_datastore_keys(
+            &self.final_state.read().ledger,
+            &self.active_history.read().0,
+            addr);
+
+        final_keys.append(&mut candidate_keys);
+        added_keys.append(&mut final_keys);
+        Some(added_keys
+            .iter()
+            .cloned()
+            .collect())
+    }
```

Comment:
What I would recommend:
```rust
pub fn get_datastore_keys(&self, addr: &Address) -> Option<BTreeSet<Vec<u8>>> {  // note that I renamed the function to keep consistency. I also changed the return format to use a btreeset to keep determinism in data ordering.
    // here, get the final keys from the final ledger
    let mut keys: Option<BTreeSet<Vec<u8>>> = self.final_state.read().get_datastore_keys(addr);  // TODO change this function so that it returns an Option, especially None if the ledger entry is absent !!!  Also make sure it returns a btreeset for ordering determinism

    // here, traverse the history from oldest to newest with added_changes at the end, applying additions and deletions
    let active_history = active_history.read();
    let changes_iterator = active_history.0.iter().map(|item| item.state_changes.ledger_changes).chain(std::iter::once(&self.added_changes)); 
    for ledger_changes in changes_iterator {
        match ledger_changes.get(addr) {
            // address absent from the changes
            None => (),

            // address ledger entry being reset to an absolute new list of keys
            Some(SetUpdateOrDelete::Set(new_ledger_entry)) => {
                keys = Some(new_ledger_entry.datastore.keys().cloned().collect());
            }

            // address ledger entry being updated
            Some(SetUpdateOrDelete::Update(entry_updates)) => {
                let ref_keys = keys.get_or_insert_default();
                for (ds_key, ds_update) in &entry_updates.datastore {
                    match ds_update {
                        SetOrDelete::Set(_) => ref_keys.insert(ds_key.clone()),
                        SetOrDelete::Delete => ref_keys.remove(ds_key),
                    };
                }
            }

            // address ledger entry being deleted
            Some(SetUpdateOrDelete::Delete) => {
                keys = None;
            }
        }
    }
    
    keys
}
```